pkg/applicationserver/metadata/location_registry.go

```diff
@@ -0,0 +1,246 @@
+}
+
+var (
+	endDeviceLocationPath      = []string{"locations"}
+	endDeviceLocationFieldMask = &pbtypes.FieldMask{Paths: endDeviceLocationPath}
+)
+
+type clusterEndDeviceLocationRegistry struct {
+	ClusterPeerAccess
+	timeout time.Duration
+}
+
+// Get implements EndDeviceLocationRegistry.
+func (c clusterEndDeviceLocationRegistry) Get(ctx context.Context, ids ttnpb.EndDeviceIdentifiers) (map[string]*ttnpb.Location, error) {
+	cc, err := c.GetPeerConn(ctx, ttnpb.ClusterRole_ENTITY_REGISTRY, nil)
+	if err != nil {
+		return nil, err
+	}
+	cl := ttnpb.NewEndDeviceRegistryClient(cc)
+	ctx, cancel := context.WithTimeout(ctx, c.timeout)
+	defer cancel()
+	dev, err := cl.Get(ctx, &ttnpb.GetEndDeviceRequest{
+		EndDeviceIdentifiers: ids,
+		FieldMask:            endDeviceLocationFieldMask,
+	}, c.WithClusterAuth())
+	if err != nil {
+		return nil, err
+	}
+	return dev.Locations, nil
+}
+
+// Merge implements EndDeviceLocationRegistry.
+func (c clusterEndDeviceLocationRegistry) Merge(ctx context.Context, ids ttnpb.EndDeviceIdentifiers, update map[string]*ttnpb.Location) (map[string]*ttnpb.Location, error) {
+	cc, err := c.GetPeerConn(ctx, ttnpb.ClusterRole_ENTITY_REGISTRY, nil)
+	if err != nil {
+		return nil, err
+	}
+	cl := ttnpb.NewEndDeviceRegistryClient(cc)
+	ctx, cancel := context.WithTimeout(ctx, c.timeout)
+	defer cancel()
+	dev, err := cl.Get(ctx, &ttnpb.GetEndDeviceRequest{
+		EndDeviceIdentifiers: ids,
+		FieldMask:            endDeviceLocationFieldMask,
+	}, c.WithClusterAuth())
+	if err != nil {
+		return nil, err
+	}
+	if len(update) == 0 {
+		return dev.Locations, nil
+	}
+	if dev.Locations == nil {
+		dev.Locations = make(map[string]*ttnpb.Location, len(update))
+	}
+	for k, l := range update {
+		dev.Locations[k] = l
+	}
+	_, err = cl.Update(ctx, &ttnpb.UpdateEndDeviceRequest{
+		EndDevice: ttnpb.EndDevice{
+			EndDeviceIdentifiers: ids,
+			Locations:            dev.Locations,
+		},
+		FieldMask: endDeviceLocationFieldMask,
+	}, c.WithClusterAuth())
+	if err != nil {
+		return nil, err
+	}
+	return dev.Locations, nil
+}
+
+// NewClusterEndDeviceLocationRegistry returns an EndDeviceLocationRegistry connected to the Entity Registry.
+func NewClusterEndDeviceLocationRegistry(cluster ClusterPeerAccess, timeout time.Duration) EndDeviceLocationRegistry {
+	return &clusterEndDeviceLocationRegistry{
+		ClusterPeerAccess: cluster,
+		timeout:           timeout,
+	}
+}
+
+type cachedEndDeviceLocationRegistry struct {
+	registry EndDeviceLocationRegistry
+	cache    EndDeviceLocationCache
+
+	softTTL  time.Duration
+	hardTTL  time.Duration
+	errorTTL time.Duration
+
+	replicationPool workerpool.WorkerPool
+}
+
+// Get implements EndDeviceLocationRegistry.
+func (c *cachedEndDeviceLocationRegistry) Get(ctx context.Context, ids ttnpb.EndDeviceIdentifiers) (map[string]*ttnpb.Location, error) {
+	locations, ttl, err := c.cache.Get(ctx, ids)
+	switch {
+	case err != nil && !errors.IsNotFound(err):
+		return nil, err
+	case err == nil:
```

Comment:
What if `errors.IsNotFound(err)`?

Reply:
We don't enter the `switch` and trigger an asynchronous replication. The result of the `Get` is empty (no locations).

Reply:
So if `errors.IsNotFound()` then one of the return values is still used; it could be nil or non-nil, or even have dirty values, depending on the implementation of `Get()`. We usually don't rely on other return values if `err` isn't `nil`.

Reply:
Added a `nil` setup.